enhancements/olm/max-openshift-versions-for-operators.md

```diff
@@ -0,0 +1,276 @@
+
+### Test Plan
+
+#### Proposed Unit Tests
+
+- Logic used to determine upgrade safety should be thoroughly tested
+
+#### Proposed E2E Tests
+
+- The cluster upgrade is not blocked by CSVs that do not report a `maxOpenShiftVersion`.
+- The cluster upgrade is not blocked by CSVs that report an invalid `maxOpenShiftVersion`.
+- The cluster upgrade is not blocked by CSVs that report a `maxOpenShiftVersion` greater than the current OpenShift version.
+- The cluster upgrade is blocked by CSVs that report a `maxOpenShiftVersion` less than or equal to the current OpenShift version..
+
+For operators that specify a `maxOpenShiftVersion`:
+
+- The operator can be installed on a Kubernetes cluster.
+- The operator can be installed on OpenShift clusters whose version is less than or equal to the `maxOpenShiftVersion` specified by the operator.
+- The operator can be installed on OpenShift clusters whose version is greater than the `maxOpenShiftVersion` specified by the operator.
+
+For operators that do not specify a `maxOpenShiftVersion`:
+
+- The operator continues to be installed on both OpenShift and vanilla Kubernetes clusters.
+
+### Graduation Criteria
+
+The goal of this enhancement is to provide Cluster Admins some level of assurance that service made available via OLM Managed Operators continue to remain available on a targeted OpenShift Cluster Version.
+
+This feature will initially be introduced as a Generally Available Feature, but upgrade compatibility checks may be added overtime.
+
+Proposed GA Features:
+
+- OLM determines upgrade safety based on CSV annotations.
+- OLM prevents operators from being installed on OpenShift clusters whose versions are greater than than the specified `maxOpenShiftVersion`.
+- Console highlights upgrade safety in the UI.
+- User facing documentation available at olm.operatorframework.io.
+- Comprehensive unit tests exist.
+- Comprehensive e2e tests exist.
+
+#### Removing a deprecated feature
+
+Not applicable.
+
+#### Upgrade / Downgrade Strategy
+
+Not applicable.
+
+#### Version Skew Strategy
+
+The proposed feature should only require integration with the CVO component and OLM Managed Operators.
+
+## Implementation History
+
+- Initial enhancement proposal created.
+
+## Drawbacks
+
+- OpenShift users might be upset when a Minor OpenShift Upgrade is block by OLM if they are unfamiliar with this feature.
+- There are instances where an operator might not be supported on the targeted version that the operator is being upgraded to. If the customer is unable to remove the existing operator, this may place them in a position where they must choose between upgrading the cluster or removing the operator. This is arguably better than upgrading the cluster only to find out that a core service is no longer available on cluster, but I suspect that multiple tickets will be opened against OLM and the operators it installs when the customer finds themselves unable to upgrade.
+
+## Possible Future Work
```

Comment:
another bit of future work that's at least related to this is providing the ability to override the install requirement.